src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs

```diff
@@ -455,8 +455,9 @@ private void NotifyFileSystemEventArgs(WatcherChangeTypes changeType, string nam
                     Debug.Fail("Unknown FileSystemEvent change type!  Value: " + changeType);
                     break;
             }
-
-            if (handler != null && MatchPattern(name))
+            
+            bool passedPatternMatch = (string.IsNullOrEmpty(name) ? MatchPattern(_directory) : MatchPattern(name));
+            if (handler != null && passedPatternMatch)
```

Comment:
This is going to invoke MatchPattern even if handler is null.  Will it be common for handler to be null?  You could instead write it like:
```C#
if (handler != null && MatchPattern(string.IsNullOrEmpty(name) ? _directory : name))
{
    handler(...);
}
```